pkg/ssh/ssh.go

```diff
@@ -379,38 +366,42 @@ func (l *SSHTunnelList) Dial(ctx context.Context, net, addr string) (net.Conn, e
 func (l *SSHTunnelList) pickTunnel(addr string) (tunnel, error) {
 	l.tunnelsLock.Lock()
 	defer l.tunnelsLock.Unlock()
-	if len(l.entries) == 0 {
-		return nil, fmt.Errorf("No SSH tunnels currently open. Were the targets able to accept an ssh-key for user %q?", l.user)
+	if len(l.tunnels) == 0 {
+		return nil, fmt.Errorf("no SSH tunnels currently open, were the targets able to accept an ssh-key for user %q", l.user)
 	}
 	// Prefer same tunnel as kubelet
-	// TODO: Change l.entries to a map of address->tunnel
-	for _, entry := range l.entries {
-		if entry.Address == addr {
-			return entry.Tunnel, nil
-		}
+	if t, ok := l.tunnels[addr]; ok {
+		return t, nil
 	}
 	klog.Warningf("SSH tunnel not found for address %q, picking random node", addr)
-	n := mathrand.Intn(len(l.entries))
-	return l.entries[n].Tunnel, nil
+	n := mathrand.Intn(len(l.tunnels))
```

Comment:
This is kind of gross, but I don't see a better way.

Reply:
yep. I don't find a way to make it more beautiful yet.

Reply:
If we keep a slice of addresses, then we can use the nth address for the key to the tunnels map. Its not only a little cleaner but I believe pickTunnel() is the most commonly called method here. I think a slice index and map lookup by key will be faster than iterating the map for the nth item for any significant number of tunnels.

Reply:
Until I see evidence that pickTunnel is a bottleneck, I'm going to favor readability. I was going to suggest that we could iterate and return the first element encountered. This should be random if the map's hash is good:

https://github.com/golang/go/blob/31c4e099158b0e4999c05ee4daf08531f6640ad4/src/runtime/map.go#L831-L837

Apparently this isn't part of the language spec though. "The iteration order over maps is not specified and is not guaranteed to be the same from one iteration to the next."

Reply:
I agree w/ Mike. Any call here is gonna be dominated by the subsequent `Dial()` and other I/O. The iteration feels gross, but I also couldn't come up w/ a cleaner way that keeps the "probably mostly random" behavior.

Reply:
Map iteration order is **already randomized**, just always take the first one if you want to do this!

Reply:
@lavalamp I mentioned that in my comment, but it's not part of the language spec.